staging/src/k8s.io/client-go/discovery/discovery_client_test.go

```diff
@@ -26,7 +26,7 @@ import (
 	"testing"
 	"time"
 
-	"github.com/gogo/protobuf/proto"
+	"github.com/golang/protobuf/proto"
```

Comment:
Just replace gogo/protobuf with golang/protobuf here. 
@beorn7

Reply:
I'm a bit confused by this… is this an example of switching all usage, or was this specific location causing a problem?

Reply:
> is this an example of switching all usage

Not really. 
The `gogo/protobuf/proto.Marshal()` can not marshal the  `openapi_v2.Document` anymore if we update `github.com/golang/protobuf` to v1.4.0, because `openapi_v2.Document` depends on `github.com/golang/protobuf`'s lower version.

I've tried update `github.com/googleapis/gnostic` to a newer pseudo version(on master), but it doesn't work, I didn't dig on it.

Reply:
One more thing, I noticed the openapi_v2.Document is Unmarshalled by the `github.com/golang/protobuf/proto` package:
https://github.com/kubernetes/kubernetes/blob/9380cb1b137b05ff12cf5a2f733653be82ff7657/staging/src/k8s.io/client-go/discovery/discovery_client.go#L437

I think it'd be better to Marshal the openapi_v2.Document  by the same package.

Reply:
> The gogo/protobuf/proto.Marshal() can not marshal the openapi_v2.Document anymore if we update github.com/golang/protobuf to v1.4.0, because openapi_v2.Document depends on github.com/golang/protobuf's lower version.

Do you have more details about this? This PR doesn't modify github.com/gogo/protobuf/proto or github.com/googleapis/gnostic/openapiv2 ... what changed that caused github.com/gogo/protobuf to break? we use github.com/gogo/protobuf in many other places, are any of them affected?

Reply:
According to the 1.4.0 release notes:

> This release maintains backwards compatibility with previous releases of this module. Any observable changes in behavior are to fix bugs, change unspecified behavior, or to make behavior more compliant with the protobuf specification.

So the breakage you describe sounds unexpected

Reply:
There are 3 commits in this PR.
The first commit focus on update client_golang, and the second commit focus on update transitive dependencies.

Then I met a panic as I mentioned [before](https://github.com/kubernetes/kubernetes/pull/90582#issuecomment-621172241) where I also put a folded `unit test panic back trace` too.

The panic message is `not implemented`, which dumped from golang/protobuf@v1.4.0:
https://github.com/golang/protobuf/blob/1b794fe86dd6a0c7c52ae69b5c9cb0aeedc52afa/proto/deprecated.go#L91

Even though the golang/protobuf@v1.4.0 maintains backward compatibility with previous releases, I guess that means all exported things are still keep exported, but the implementation of [type InternalMessageInfo struct{}](https://github.com/golang/protobuf/blob/1b794fe86dd6a0c7c52ae69b5c9cb0aeedc52afa/proto/deprecated.go#L86) has been changed, all methods of this struct have been disabled by raising a panic.
```
func (*InternalMessageInfo) Size(Message) int { panic("not implemented") }
```

Let me try to explain why this panic is triggered by https://github.com/kubernetes/kubernetes/blob/863ce9726e0aa617f03332f42ff3cebb7abe8fc0/staging/src/k8s.io/client-go/discovery/discovery_client_test.go#L474
When the function `github.com/gogo/protobuf/proto.Marshal(*openapi_v2.Document)` marshal an openapi document, i will call [github.com/googleapis/gnostic/openapiv2.Document->XXX_Size()](https://github.com/googleapis/gnostic/blob/896953e6749863beec38e27029c804e88c3144b8/openapiv2/OpenAPIv2.pb.go#L550).

The thing is `XXX_Size()` depends on `golang/protobuf/proto.InternalMessageInfo`.
```
var xxx_messageInfo_Document proto.InternalMessageInfo
func (m *Document) XXX_Size() int {
	return xxx_messageInfo_Document.Size(m)
}
```
Even though we didn't change `github.com/gogo/protobuf` and `github.com/googleapis/gnostic`, but we can't use `golang/protobuf/proto.InternalMessageInfo` anymore as it has been deprecated(panic).

So, the third commits just simply changed the marshal function from `github.com/gogo/protobuf` to `github.com/golang/protobuf`.
  
@liggitt

Reply:
> we use github.com/gogo/protobuf in many other places, are any of them affected?

There are only two places used gogo/protobuf/proto.Marshal() and I tried to fix them in the third commit.

I don't know if there are any potential risks. I didn't mean to remove gogo/protobuf in this PR, but after fixing the panic, it seems client-go doesn't need it anymore.

Reply:
is the issue that we have stale generated code referencing proto.InternalMessageInfo?

There are many instances of that still in the codebase:

```
vendor/google.golang.org/grpc/health/grpc_health_v1/health.pb.go:	proto "github.com/golang/protobuf/proto"
vendor/google.golang.org/grpc/health/grpc_health_v1/health.pb.go:var xxx_messageInfo_HealthCheckRequest proto.InternalMessageInfo
--
vendor/google.golang.org/grpc/binarylog/grpc_binarylog_v1/binarylog.pb.go:import proto "github.com/golang/protobuf/proto"
vendor/google.golang.org/grpc/binarylog/grpc_binarylog_v1/binarylog.pb.go:var xxx_messageInfo_GrpcLogEntry proto.InternalMessageInfo
--
vendor/google.golang.org/appengine/internal/app_identity/app_identity_service.pb.go:import proto "github.com/golang/protobuf/proto"
vendor/google.golang.org/appengine/internal/app_identity/app_identity_service.pb.go:var xxx_messageInfo_AppIdentityServiceError proto.InternalMessageInfo
--
vendor/google.golang.org/appengine/internal/remote_api/remote_api.pb.go:import proto "github.com/golang/protobuf/proto"
vendor/google.golang.org/appengine/internal/remote_api/remote_api.pb.go:var xxx_messageInfo_Request proto.InternalMessageInfo
--
vendor/google.golang.org/appengine/internal/datastore/datastore_v3.pb.go:import proto "github.com/golang/protobuf/proto"
vendor/google.golang.org/appengine/internal/datastore/datastore_v3.pb.go:var xxx_messageInfo_Action proto.InternalMessageInfo
--
vendor/google.golang.org/appengine/internal/log/log_service.pb.go:import proto "github.com/golang/protobuf/proto"
vendor/google.golang.org/appengine/internal/log/log_service.pb.go:var xxx_messageInfo_LogServiceError proto.InternalMessageInfo
--
vendor/google.golang.org/appengine/internal/modules/modules_service.pb.go:import proto "github.com/golang/protobuf/proto"
vendor/google.golang.org/appengine/internal/modules/modules_service.pb.go:var xxx_messageInfo_ModulesServiceError proto.InternalMessageInfo
--
vendor/google.golang.org/appengine/internal/urlfetch/urlfetch_service.pb.go:import proto "github.com/golang/protobuf/proto"
vendor/google.golang.org/appengine/internal/urlfetch/urlfetch_service.pb.go:var xxx_messageInfo_URLFetchServiceError proto.InternalMessageInfo
--
vendor/google.golang.org/appengine/internal/base/api_base.pb.go:import proto "github.com/golang/protobuf/proto"
vendor/google.golang.org/appengine/internal/base/api_base.pb.go:var xxx_messageInfo_StringProto proto.InternalMessageInfo
--
vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/proto/client/client.pb.go:	proto "github.com/golang/protobuf/proto"
vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/proto/client/client.pb.go:var xxx_messageInfo_Packet proto.InternalMessageInfo
--
vendor/github.com/grpc-ecosystem/grpc-gateway/internal/stream_chunk.pb.go:import proto "github.com/golang/protobuf/proto"
vendor/github.com/grpc-ecosystem/grpc-gateway/internal/stream_chunk.pb.go:var xxx_messageInfo_StreamError proto.InternalMessageInfo
--
vendor/github.com/googleapis/gnostic/extensions/extension.pb.go:	proto "github.com/golang/protobuf/proto"
vendor/github.com/googleapis/gnostic/extensions/extension.pb.go:var xxx_messageInfo_Version proto.InternalMessageInfo
--
vendor/github.com/googleapis/gnostic/openapiv2/OpenAPIv2.pb.go:	proto "github.com/golang/protobuf/proto"
vendor/github.com/googleapis/gnostic/openapiv2/OpenAPIv2.pb.go:var xxx_messageInfo_AdditionalPropertiesItem proto.InternalMessageInfo
--
vendor/github.com/container-storage-interface/spec/lib/go/csi/csi.pb.go:	proto "github.com/golang/protobuf/proto"
vendor/github.com/container-storage-interface/spec/lib/go/csi/csi.pb.go:var xxx_messageInfo_GetPluginInfoRequest proto.InternalMessageInfo
```

are all of those potentially panic-inducing now?

Reply:
> are all of those potentially panic-inducing now?

Yes! Awesome! I didn't check the generated code, and I'm glad you point it out.

Do you have any ideas about how to solve this?
IMO, given [github.com/golang/protobuf has been superseded by the google.golang.org/protobuf module](https://github.com/golang/protobuf#go-support-for-protocol-buffers), 
And, the recommendation from  github.com/golang/protobuf as follows:
> Versions v1.4 and later of github.com/golang/protobuf are implemented in terms of google.golang.org/protobuf. Programs which use both modules must use at least version v1.4 of this one.

I didn't check every module you list above, but I did found some of them using `protoc-gen-go`'s version lower than v1.4. such as:
https://github.com/googleapis/gnostic/blob/428ada5e5547af95eb370db4c0528778062dfbc1/openapiv2/OpenAPIv2.pb.go#L17-L21

I think, the `*.pb.go` in the vendor modules, should be regenerated by protoc-gen-go@v1.4.0+, or regernerated by the replacement [protogen](https://github.com/protocolbuffers/protobuf-go/tree/master/compiler/protogen). How do you say?

By the way, I still have a question in my mind.
I can't figure out why no panic found In this PR's CI jobs? Does that mean some checkpoint missing?

Reply:
> I think, the `*.pb.go` in the vendor modules, should be regenerated by protoc-gen-go@v1.4.0+, or regernerated by the replacement [protogen](https://github.com/protocolbuffers/protobuf-go/tree/master/compiler/protogen). How do you say?

We do not modify the content of the `vendor` dir, so if changes are required in our dependencies, that must start as a PR to the upstream project(s), then an update in k8s.io/kubernetes once that is merged and released

> I can't figure out why no panic found In this PR's CI jobs? Does that mean some checkpoint missing?

We don't run all the unit tests of all our dependencies. I expect there would be more detailed coverage in those that could expose some of these issues.

Reply:
See my comment here: https://github.com/golang/protobuf/issues/1126#issuecomment-628375275

If it helps, we can cut a `github.com/golang/protobuf@v1.4.2` release that adds back minimal support for `InternalMessageInfo`. The fundamental problem is that the `gogo/protobuf` implementation violates documented guidelines to **not** call certain API endpoints.

Reply:
Awesome! It works with your patch, after I discard the third commit and run following commands, unit test goes all well.
```
# hack/pin-dependency.sh github.com/golang/protobuf f183159caab1d3a7dbf85306153a71aed61e9d8d
# hack/update-vendor.sh
# go test ./vendor/k8s.io/client-go/discovery/
ok  	k8s.io/client-go/discovery	0.033s
```

Thanks, @dsnet for your kindly support. 

So, if it acceptable for you to go with `github.com/golang/protobuf@v1.4.2`?